### PR TITLE
feat: source subscribe

### PR DESCRIPTION
**Pull request by capJavert**
### Describe what this PR does
- [x] add option to subscribe to source
- [x] api implementation https://github.com/dailydotdev/daily-api/pull/1739
- [x] from source page
- [x] from feed card
- [x] from post page/modal

Available on [preview](https://preview.app.daily.dev/)

TODO

- [x] pending [thread resolution](https://dailydotdev.slack.com/archives/C02JAUF8HJL/p1709800951729079) for subscribe in post sidebar

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change | open signup | extra: { trigger: source subscribe } |

### On those affected packages:
- [x] Have you done sanity checks in the webapp?
- [x] Have you done sanity checks in the extension?
- [x] Does this not break anything in companion?

### Did you test the modified components media queries?
- [x] MobileL (420px)
- [x] Tablet (656px)
- [x] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

AS-154 #done

### CODE REVIEW

**Review by rebelchris — APPROVED** [review]
Looks ok, just make sure to check if we need tracking events 🙏

File: packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  params: UseNotificationPreferenceProps['params'][0] | undefined;
+};
+
+export type UseNotificationPreferenceToggle = {
+  isSubscribed: boolean;
+  isReady: boolean;
+  onToggle: () => Promise<{ isSubscribed: boolean }>;
+};
+
+export const useNotificationPreferenceToggle = ({
```
[the remark this answers is not in the document]
Comment (capJavert) [reply packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts:20]: We have `useNotificationPreference` but it requires lot of boilerplate code to setup this kind of toggle feature. Look at `CollectionSubscribeButton` or `SquadNotificationsModal`. This one is lean to managed toggle of notification type subscription.

[the remark this answers is not in the document]
Comment (capJavert) [reply packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts:20]: This one can be used in other places like `CollectionSubscribeButton` but did not want to refactor in this one to minimize changes. I can update in separate PR or add a task.

File: packages/shared/src/components/PostOptionsMenu.tsx [hunk 2]
```diff
@@ -334,6 +361,20 @@ export default function PostOptionsMenu({
...
+  if (isLoggedIn && isSourceSubscribeV1 && !isSourceBlocked) {
+    postOptions.push({
+      icon: (
+        <MenuIcon
+          Icon={sourceSubscribe.isSubscribed ? BellSubscribedIcon : BellIcon}
+        />
+      ),
+      label: sourceSubscribe.isSubscribed
+        ? `Unsubscribe from ${post?.source?.name}`
+        : `Subscribe to ${post?.source?.name}`,
```
Comment (kopancek) [inline packages/shared/src/components/PostOptionsMenu.tsx:373]: The suffix with name seems to be the same in both cases, we could do the string literal in a different way:
```suggestion
      label: `${sourceSubscribe.isSubscribed ? 'Unsubscribe from' : 'Subscribe to'} ${post?.source?.name}`,
```

File: packages/shared/src/components/sources/SourceSubscribeButton.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    const result = await onToggle();
+
+    displayToast(
+      result.isSubscribed
+        ? '✅ You are now subscribed'
+        : '⛔️ You are now unsubscribed',
+    );
+  };
```
Comment (kopancek) [inline packages/shared/src/components/sources/SourceSubscribeButton.tsx:98]: This functionality is the same for this button as for the options menu. Maybe we can extract it out and use in both places?

File: packages/shared/src/components/widgets/PostUsersHighlights.tsx [hunk 4]
```diff
@@ -88,12 +94,25 @@ const UserHighlight = (props: SourceAuthorProps) => {
     reputation,
   } = props;
   const Icon = getUserIcon(userType);
-  const userTypeNotSource = userType !== 'source';
+  const isUserTypeSource = userType === 'source';
```
Comment (kopancek) [inline packages/shared/src/components/widgets/PostUsersHighlights.tsx:97]: Should we add an enum for userType and use it instead of a string value here?

File: packages/shared/src/components/widgets/PostUsersHighlights.tsx [hunk 5]
```diff
@@ -116,30 +135,47 @@ const UserHighlight = (props: SourceAuthorProps) => {
...
-          </div>
+          )}
+          {!isUserTypeSource && (
+            <div className="flex">
+              <ProfileLink className="font-bold typo-callout" href={permalink}>
+                {name}
+              </ProfileLink>
+              <ReputationUserBadge user={{ reputation }} />
+            </div>
+          )}
```
Comment (kopancek) [inline packages/shared/src/components/widgets/PostUsersHighlights.tsx:161]: I think a ternary would be better here, since we are basically reversing the condition:
```jsx
          {isUserTypeSource ? (
            <ProfileLink className="font-bold typo-callout" href={permalink}>	            <ProfileLink
              className={classNames('!block truncate font-bold typo-callout')}
              href={permalink}
            >
              {name}	
            </ProfileLink>
          ) : (
            <div className="flex">
              <ProfileLink className="font-bold typo-callout" href={permalink}>
                {name}
              </ProfileLink>
              <ReputationUserBadge user={{ reputation }} />
            </div>
          )}
```

File: packages/shared/src/components/withExperiment.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import React, { ComponentType, ReactElement } from 'react';
+import { JSONValue, WidenPrimitives } from '@growthbook/growthbook';
+import { useFeature } from './GrowthBookProvider';
+import { Feature } from '../lib/featureManagement';
+
+export const withExperiment = <Props, TFeature extends JSONValue>(
```
[the remark this answers is not in the document]
Comment (kopancek) [reply packages/shared/src/components/withExperiment.tsx:6]: I think this is a great idea ❤️

File: packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      checkHasStatusPreference(
+        item,
+        params?.notificationType,
+        params?.referenceId,
+        [NotificationPreferenceStatus.Subscribed],
+      ),
+    );
+  }, [preferences, params?.notificationType, params?.referenceId]);
+
+  const { mutateAsync: onToggle } = useMutation({
```
Comment (kopancek) [inline packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts:43]: Can we call it something else than `onToggle`? E.g. even simple `toggle` would probably be better, since we are actually mutating the state here, not just reacting to toggle event in the UI.

File: packages/webapp/pages/sources/[source].tsx [hunk 8]
```diff
@@ -118,15 +120,13 @@ const SourcePage = ({ source }: SourcePageProps): ReactElement => {
...
           aria-label={unfollowingSource ? 'Follow' : 'Block'}
         />
-        <Button
-          className="hidden laptop:flex"
-          variant={ButtonVariant.Secondary}
-          {...buttonProps}
-        >
+        {!unfollowingSource && (
+          <SourceSubscribeButton className="ml-3 laptop:mr-3" source={source} />
+        )}
```
Comment (omBratteng) [inline packages/webapp/pages/sources/[source].tsx:128]: Should this be hid behind a feature flag?

**Review by kopancek — COMMENTED** [review]
Looks great, added some non-blocking comments.

File: packages/shared/src/components/PostOptionsMenu.tsx (hunk 2 above)
Comment (capJavert) [reply packages/shared/src/components/PostOptionsMenu.tsx:373]: Yes!

File: packages/shared/src/components/widgets/PostUsersHighlights.tsx (hunk 5 above)
Comment (capJavert) [reply packages/shared/src/components/widgets/PostUsersHighlights.tsx:161]: I don't really like ternaries in JSX. Hard to follow and nest. It is also experiment so it will be cleaned up a bit when either one wins.

File: packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts (hunk 7 above)
Comment (capJavert) [reply packages/shared/src/hooks/notifications/useNotificationPreferenceToggle.ts:43]: We usually call handlers with this kind of nomenclature. Don't want to change at this time if that is ok?

File: packages/shared/src/components/widgets/PostUsersHighlights.tsx (hunk 4 above)
Comment (capJavert) [reply packages/shared/src/components/widgets/PostUsersHighlights.tsx:97]: Its a local type for a component. I can do enum yes.

File: packages/webapp/pages/sources/[source].tsx (hunk 8 above)
Comment (capJavert) [reply packages/webapp/pages/sources/[source].tsx:128]: Component is wrapped in [withExperiment](https://github.com/dailydotdev/apps/pull/2750#discussion_r1516128240) so it already is behind it.

File: packages/shared/src/components/widgets/PostUsersHighlights.tsx (hunk 5 above)
Comment (omBratteng) [reply packages/shared/src/components/widgets/PostUsersHighlights.tsx:161]: If it weren't for the extra classNames, we could've used ConditionalWrapper
```tsx
<ConditionalWrapper
  condition={!isUserTypeSource}
  wrapper={(component) => (
    <div className="flex">
      {component}
      <ReputationUserBadge user={{ reputation }} />
    </div>
  )}
>
```

Maybe it might work if `component` is `Component` with the props added? Don't know if that'll work.
Could also just add that onto the ProfileLink itself

```tsx
<ConditionalWrapper
  condition={!isUserTypeSource}
  wrapper={(component) => (
    <div className="flex">
      {component}
      <ReputationUserBadge user={{ reputation }} />
    </div>
  )}
>
  <ProfileLink
    className={classNames(
      "font-bold typo-callout",
      isUserTypeSource && "!block truncate")}
    href={permalink}
  >
    {name}
  </ProfileLink>
</ConditionalWrapper>
```

File: packages/webapp/pages/sources/[source].tsx (hunk 8 above)
Comment (omBratteng) [reply packages/webapp/pages/sources/[source].tsx:128]: Ah yes! Didn't see that part 😬

File: packages/shared/src/components/sources/SourceSubscribeButton.tsx (hunk 3 above)
Comment (capJavert) [reply packages/shared/src/components/sources/SourceSubscribeButton.tsx:98]: Any ideas how you would do it?

File: packages/shared/src/components/widgets/PostUsersHighlights.tsx (hunk 5 above)
Comment (capJavert) [reply packages/shared/src/components/widgets/PostUsersHighlights.tsx:166]: Did not manage truncate to work with reputation badge. I don't have the nerve to play with CSS that much.

Comment (omBratteng) [reply packages/shared/src/components/widgets/PostUsersHighlights.tsx:166]: truncating is... yeah... I've done a few PRs this week with them 😿

**Review by omBratteng — APPROVED** [review]

File: packages/shared/src/components/sources/SourceSubscribeButton.tsx (hunk 3 above)
Comment (capJavert) [reply packages/shared/src/components/sources/SourceSubscribeButton.tsx:98]: I'll create a hook to manage it for now.

Comment (capJavert) [reply packages/shared/src/components/sources/SourceSubscribeButton.tsx]: See: 9f96b376acb9bce561b75f14339166cd875444dd

**Review by omBratteng — APPROVED** [review]

**Review by nensidosari — APPROVED** [review]
Looks good 🎉
